src/unix/timer.c

```diff
@@ -71,7 +72,7 @@ int uv_timer_start(uv_timer_t* handle,
   if (uv__is_active(handle))
     uv_timer_stop(handle);
 
-  clamped_timeout = handle->loop->time + timeout;
+  clamped_timeout = (handle->timeout <= 0 ? handle->loop->time : handle->timeout) + timeout;
```

Comment:
That's why I mentioned uv_update_time().  Libuv will update the time for you at fixed points during the event loop cycle but not on every call to uv_timer_start(), that's too expensive.